test/hotspot/jtreg/compiler/codecache/CheckLargePages.java

```diff
@@ -44,35 +44,83 @@
 import java.util.List;
 
 public class CheckLargePages {
-    private final static WhiteBox WHITE_BOX = WhiteBox.getWhiteBox();
+    private final static long LP_1G = 1024 * 1024 * 1024;
+    private final static boolean LARGE_PAGES_ENABLED;
+    private final static long LARGE_PAGE_SIZE;
 
+    static {
+        WhiteBox whiteBox = WhiteBox.getWhiteBox();
+        LARGE_PAGES_ENABLED = whiteBox.getBooleanVMFlag("UseLargePages");
+        LARGE_PAGE_SIZE = (whiteBox.getBooleanVMFlag("UseLargePages")) ? whiteBox.getVMLargePageSize() : 0;
+    }
+
+    private static boolean isLargePageSizeEqual(long size) {
+        return LARGE_PAGE_SIZE == size;
+    }
+
+    private static void testSegmented2GbCodeCacheWith1GbPage() throws Exception {
+        ProcessBuilder pb = ProcessTools.createLimitedTestJavaProcessBuilder(
+                "-XX:+UseLargePages",
+                "-XX:+SegmentedCodeCache",
+                "-XX:InitialCodeCacheSize=2g",
+                "-XX:ReservedCodeCacheSize=2g",
+                "-XX:LargePageSizeInBytes=1g",
+                "-Xlog:pagesize=info",
+                "-version");
+        OutputAnalyzer out = new OutputAnalyzer(pb.start());
+        out.shouldMatch("Code cache size too small for \\S* pages\\. Reverting to smaller page size \\((\\S*)\\)\\.");
+        out.shouldHaveExitValue(0);
```

Comment:
I don't get this.

The new version (which I like, btw) now assumes min_pages=1 if InitialCCS == ReservedCCS. Should this not then work? I would have expected to get a code cache with 2x1G pages. I would not have expected a "code cache size too small" warning.

Reply:
> The new version (which I like, btw) now assumes min_pages=1 if InitialCCS == ReservedCCS.

It is only for the non-segmented CodeCache. Here we explicitly specify to use the segmented CodeCache. As such, we need to split 2G among three code heaps. InitialCCS is tricky in such a case, with the segmented CodeCache, the initial size of a code heap is the min of InitialCCS and the code heap size. We want the 2G segmented CodeCache to have 8 pages at least which is not possible with 1G pages. The number 8 is hard-coded. 

I think calculating the minimum number of pages is not easy for the segmented CodeCache.
We have:
- InitialCodeCacheSize
- ReservedCodeCacheSize
- LargePageSizeInBytes
- NonNMethodCodeHeapSize
- ProfiledCodeHeapSize
- NonProfiledCodeHeapSize

It's too many parameters for users.

A possible configuration: one 1G page for the non-nmethod code heap and the profiled code heap, another 1G page for the non-profiled code heap.

BTW, this is the original first test.

Reply:
Okay, I see.

We should probably allow the segmented code cache, in its entirety, to live on one or a few super large pages if we desire to; but I guess that would be too much changes for now. Not for this RFE.